File: iosApp/iosApp/Features/Home/MoonDisc.swift
import SwiftUI

/// A real moon-phase disc: a lit gold sphere with the shaded portion drawn from the true
/// illumination fraction. The terminator is a proper half-ellipse (not a flat symbol), so
/// crescents and gibbous phases read correctly. Waxing lights the right limb, waning the left.
struct MoonDisc: View {
    var illumination: Double        // 0 (new) .. 1 (full)
    var waxing: Bool
    var lit: Color = Color(hex: "EAD59B")
    var shadow: Color = Color(hex: "20152A")
    var glow: Bool = true

    var body: some View {
        Canvas { ctx, size in
            let r = min(size.width, size.height) / 2
            let c = CGPoint(x: size.width / 2, y: size.height / 2)
            let disc = Path(ellipseIn: CGRect(x: c.x - r, y: c.y - r, width: 2 * r, height: 2 * r))

            // Soft sphere shading on the lit base.
            ctx.fill(disc, with: .radialGradient(
                Gradient(colors: [Color(hex: "FFF6E0"), lit, Color(hex: "C9A24A")]),
                center: CGPoint(x: c.x - r * 0.28, y: c.y - r * 0.3),
                startRadius: 1, endRadius: r * 1.15))

            // Overlay the shadow region.
            let k = max(0, min(1, illumination))
            ctx.fill(shadowPath(c: c, r: r, k: k, waxing: waxing),
                     with: .color(shadow.opacity(0.9)))
            ctx.stroke(disc, with: .color(Color(hex: "B48C32").opacity(0.35)), lineWidth: 0.75)
        }
        .background(
            glow ? Circle().fill(lit).blur(radius: 8).opacity(0.35) : nil
        )
        .aspectRatio(1, contentMode: .fit)
    }

    /// The unlit region: the dark limb semicircle plus/minus the terminator ellipse.
    private func shadowPath(c: CGPoint, r: CGFloat, k: Double, waxing: Bool) -> Path {
        let kappa: CGFloat = 0.5523
        // Terminator horizontal offset at mid-height; +r at new, -r at full.
        let a = CGFloat(1 - 2 * k) * r * (waxing ? 1 : -1)
        let top = CGPoint(x: c.x, y: c.y - r)

        var p = Path()
        p.move(to: top)
        // Dark limb: the semicircle on the shadow side (left if waxing).
        p.addArc(center: c, radius: r,
                 startAngle: .degrees(-90), endAngle: .degrees(90),
                 clockwise: waxing)
        // Terminator ellipse back to top (two cubic quarter-ellipse segments).
        p.addCurve(to: CGPoint(x: c.x + a, y: c.y),
                   control1: CGPoint(x: c.x + a * kappa, y: c.y + r),
                   control2: CGPoint(x: c.x + a, y: c.y + r * kappa))
        p.addCurve(to: top,
                   control1: CGPoint(x: c.x + a, y: c.y - r * kappa),
                   control2: CGPoint(x: c.x + a * kappa, y: c.y - r))
        p.closeSubpath()
        return p
    }
}

File: androidApp/src/main/kotlin/app/cycluna/android/features/home/MoonDisc.kt
package app.cycluna.android.features.home

import androidx.compose.foundation.Canvas
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Rect
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.unit.dp

private val LIT = Color(0xFFEAD59B)
private val SHADOW = Color(0xFF20152A)
private val HIGHLIGHT = Color(0xFFFFF6E0)
private val LIMB = Color(0xFFC9A24A)
private val EDGE = Color(0xFFB48C32)

/**
 * A real moon-phase disc: a lit gold sphere with the shaded portion drawn from the true
 * illumination fraction. The terminator is a proper half-ellipse rather than a flat symbol,
 * so crescents and gibbous phases read correctly. Waxing lights the right limb, waning the
 * left.
 */
@Composable
fun MoonDisc(
    illumination: Double,   // 0 (new) .. 1 (full)
    waxing: Boolean,
    modifier: Modifier = Modifier,
    lit: Color = LIT,
) {
    Canvas(modifier) {
        val r = size.minDimension / 2f
        val c = Offset(size.width / 2f, size.height / 2f)

        // Soft sphere shading on the lit base, lit from the upper left.
        drawCircle(
            brush = Brush.radialGradient(
                colors = listOf(HIGHLIGHT, lit, LIMB),
                center = Offset(c.x - r * 0.28f, c.y - r * 0.3f),
                radius = r * 1.15f,
            ),
            radius = r,
            center = c,
        )

        drawPath(
            path = shadowPath(c, r, illumination.coerceIn(0.0, 1.0).toFloat(), waxing),
            color = SHADOW.copy(alpha = 0.9f),
        )

        drawCircle(
            color = EDGE.copy(alpha = 0.35f),
            radius = r,
            center = c,
            style = Stroke(width = 0.75.dp.toPx()),
        )
    }
}

/**
 * The unlit region: the dark limb's semicircle, closed by the terminator ellipse.
 *
 * The terminator is two cubic quarter-ellipses using the circle-approximation constant —
 * the same construction as the iOS twin. Getting this right is what makes a waxing gibbous
 * look like a sphere rather than a pac-man.
 */
private const val KAPPA = 0.5523f

private fun shadowPath(c: Offset, r: Float, k: Float, waxing: Boolean): Path {
    // Terminator's horizontal offset at mid-height: +r at new, -r at full.
    val a = (1f - 2f * k) * r * (if (waxing) 1f else -1f)
    val topY = c.y - r

    return Path().apply {
        moveTo(c.x, topY)
        // The dark limb: the semicircle on the shadow side (left when waxing).
        arcTo(
            rect = Rect(c.x - r, c.y - r, c.x + r, c.y + r),
            startAngleDegrees = -90f,
            sweepAngleDegrees = if (waxing) -180f else 180f,
            forceMoveTo = false,
        )
        // Terminator back up to the top.
        cubicTo(
            c.x + a * KAPPA, c.y + r,
            c.x + a, c.y + r * KAPPA,
            c.x + a, c.y,
        )
        cubicTo(
            c.x + a, c.y - r * KAPPA,
            c.x + a * KAPPA, c.y - r,
            c.x, topY,
        )
        close()
    }
}
